### PR TITLE
Api v2 + authentication packages

### PR DESCRIPTION
**Pull request by lobre**
Hello there,

This pull request is in response with the following issue:

https://github.com/juruen/rmapi/issues/54

It brings another implementation of bindings to the Remarkable Cloud API.
It aims to be self contained with well defined package boundaries. It only imports minimal external necessary packages and is not opinionated. By this, I mean it does not force you with specific logging methods for example.

These packages are libraries with explicit inputs and outputs. They can be easily imported by other packages of the project and are especially designed to be easily importable by external packages in other external projects.

In order to know how to use them, I have added as much "go doc" documentation as possible. As well, the `api.v2/client_test.go` can be used as reference to know how to use them.

As discussed in the corresponding issue, the package `auth` is decorrelated from the `api.v2` package. It means in the future another implementation of the authentication could be adapted if necessary (if Remarkable decides to bring radical changes for instance).

We discussed that we wanted the `api` package to have high abstraction public methods that can be easily used. The `api.v2` package has the following public methods exposed (defined in `api.v2/porcelain.go`:

```
func (c *Client) CreateFolder(name string, parent string) (string, error)
func (c *Client) Delete(uuid string) error
func (c *Client) Download(uuid string, w io.Writer) error
func (c *Client) Get(uuid string) (Document, error)
func (c *Client) List() ([]Document, error)
func (c *Client) Metadata(doc Document) error
func (c *Client) Upload(uuid string, name string, r io.Reader) error
func (c *Client) UploadDocument(doc Document, r io.Reader) error
```
You can have a look to the "go doc" documentation for each method. I have tried to be as idiomatic as possible. For `Download` and `Upload`, I used `io.Writer` and `io.Reader` to be able to directly download a document and return it to a `http.ResponseWriter` for example.

These packages are meant to use in combination with the `archive` package I already wrote. This way, the `api.v2` package only deals with api stuff and is not responsible for creating a correct `document.zip`. This is easier to maintain I think because each package has a single responsibility. If the format of a Remarkable archive changes, we won't have to update the `api` package.

The `auth` package defines the `TokenStore` interface and implements it with the default `FileTokenStore` struct. This will save and load tokens by default from the `$HOME/.rmapi` yaml file that already exists in the project. But another struct could implement the `TokenStore` interface as well to save tokens into a database or a HashCorp Vault instance for example.

So far, I just brought these two `api.v2` and `auth` packages. I did not refactor anything as I wanted feedback before. I really like the idea of moving all little features in specific packages that would act as kits/lego/libraries to main command programs.

In this Github repository, we could say that the `shell` is the main command program. By the way, it could be moved someday to a `cmd/shell` folder as it is rather a good practice seen in the community.
This `shell` command could be opinionated by defining a logger, implementing the real program interface with the end user and dealing with errors when encountering them. the `shell` package would import the libraries, kits as any command program could do.

To enforce this philosophy, the few packages at the root level that are not kits and that are not really usable from the outside in another context than the shell (`util`, `model`, `transport`) could be slowly removed or moved to the `cmd/shell` package (as they are only used by the `shell` package).

@juruen, I would like to have your opinion about all of this. So far, it is just me thinking about a handy architecture that would encourage other Golang open source projects around the Remarkable!

In the meantime of this refactoring, I started a rapid external project that will act as a showcase of this new api design. It is a simple web interface in Golang that allows a user to upload a PDF file to the Remarkable Cloud. It is a work in progress but it will take care of:
 - gathering the PDF file from the http request
 - creating an `archive.zip` with the `archive` package
 - using the `api.v2` package to send it to the Remarkable Cloud

The code is here: https://github.com/lobre/rmuploader

Sorry about this gigantic PR comment and happy reading if you find some time someday!
Have a good day,
Loric

### CODE REVIEW

**Comment by juruen** [comment]
@lobre brilliant work! Am super happy to see this done!

Let me have a closer look, but I really like how this is looking like!

Thank you so much!

File: auth/auth.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+	if tks.UserToken != "" && !a.Refresh {
+		return tks.UserToken, nil
+	}
+
+	if tks.DeviceToken == "" {
+		return "", errors.New("auth: nil DeviceToken, please register device")
+	}
+
+	tks.UserToken, err = renewToken(tks.DeviceToken)
```
Comment (lobre) [inline auth/auth.go]: Just notice that I could reset the `a.Refresh` flag to `false` when the token is renewed.

**Comment by lobre** [comment]
Sure don't hesitate to ask questions/post review if anything is unclear, badly documented or missing!

I can see as well that tests are failing:

https://dev.azure.com/fruruenv/fruruenv/_build/results?buildId=33

Don't know exactly how to fix because tests require a valid `.rmapi` file that could authenticate for testing the API :/. And of course it does not exist in the context of the azure pipeline.

**Comment by juruen** [comment]
> Don't know exactly how to fix because tests require a valid `.rmapi` file that could authenticate for testing the API :/. And of course it does not exist in the context of the azure pipeline.

My proposal would be to mark the test files that require a valid `.rmapi` file with:

```
// +build integration
```

And whenever you need to run them locally you can use:

```
go test -tags=integration
```

If I had infinite time, I'd probably try to create a `remarkable mock server` to use in the integration tests. It'd be cool but a bit overkill at the moment 😄

**Comment by lobre** [comment]
I did not know we could use "custom" build tags. I thought it was only working with a predefined set of OS and architectures. Seems like an interesting feature!

It could effectively be helpful in this situation. Totally agree that a mocking service would rock here :smiley:.

I am not sure about the term "integration" however. That would mean that the test file woud be taken into consideration under an "integration" context. But it feels like the opposite to me. We don't want to care about them when we are running the "integration" tests. We want them locally when authenticated with a real API account.

Maybe something among the following?
- connected
- authenticated
- withtoken
- linked

**Comment by lobre** [comment]
Hi @juruen,

I have gone with the `withauth` build tag. So the CI now ignore as wanted these tests.
I have also added the `go.sum` as it should be kept under version control.

See: https://blog.golang.org/using-go-modules
> Both go.mod and go.sum should be checked into version control.

Loric

**Comment by juruen** [comment]
> I have gone with the `withauth` build tag. So the CI now ignore as wanted these tests.
> I have also added the `go.sum` as it should be kept under version control.

Great, thank you so much for doing that!

**Pull request merged**